{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *cis*- and *trans*-QTL mapping with tensorQTL\n",
    "\n",
    "This notebook provides examples for running *cis*- and *trans*-QTL mapping with tensorQTL, using open-access data from the [GEUVADIS](https://www.ebi.ac.uk/arrayexpress/experiments/E-GEUV-1/) project.\n",
    "\n",
    "#### Requirements\n",
    "An environment configured with a GPU and ~50GB of memory.\n",
    "\n",
    "#### Test dataset\n",
    "\n",
    "*Note: these files are provided for testing/benchmarking purposes only. They do not constitute an official release from the GEUVADIS project, and no quality-control was applied.*\n",
    "\n",
    "Genotypes in PLINK and VCF format, and normalized expression data are available [here](https://personal.broadinstitute.org/francois/geuvadis/).\n",
    "\n",
    "Alternatively, to download the files required for these examples, uncomment and run the cell below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# !wget https://personal.broadinstitute.org/francois/geuvadis/GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.bed\n",
    "# !wget https://personal.broadinstitute.org/francois/geuvadis/GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.bim\n",
    "# !wget https://personal.broadinstitute.org/francois/geuvadis/GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup.fam   \n",
    "# !wget https://personal.broadinstitute.org/francois/geuvadis/GEUVADIS.445_samples.covariates.txt\n",
    "# !wget https://personal.broadinstitute.org/francois/geuvadis/GEUVADIS.445_samples.expression.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PyTorch 1.7.1\n",
      "Pandas 1.2.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:19<00:00,  6.52s/it]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import torch\n",
    "import tensorqtl\n",
    "from tensorqtl import genotypeio, cis, trans\n",
    "print('PyTorch {}'.format(torch.__version__))\n",
    "print('Pandas {}'.format(pd.__version__))\n",
    "\n",
    "# define paths to data\n",
    "plink_prefix_path = 'GEUVADIS.445_samples.GRCh38.20170504.maf01.filtered.nodup'\n",
    "expression_bed = 'GEUVADIS.445_samples.expression.bed.gz'\n",
    "covariates_file = 'GEUVADIS.445_samples.covariates.txt'\n",
    "prefix = 'GEUVADIS.445_samples'\n",
    "\n",
    "# load phenotypes and covariates\n",
    "phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)\n",
    "covariates_df = pd.read_csv(covariates_file, sep='\\t', index_col=0).T\n",
    "\n",
    "# PLINK reader for genotypes\n",
    "pr = genotypeio.PlinkReader(plink_prefix_path)\n",
    "genotype_df = pr.load_genotypes()\n",
    "variant_df = pr.bim.set_index('snp')[['chrom', 'pos']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *cis*-QTL: nominal p-values for all variant-phenotype pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 445 samples\n",
      "  * 301 phenotypes\n",
      "  * 26 covariates\n",
      "  * 13369268 variants\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr18\n",
      "    processing phenotype 301/301\n",
      "    time elapsed: 0.04 min\n",
      "    * writing output\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# map all cis-associations (results for each chromosome are written to file)\n",
    "\n",
    "# all genes\n",
    "# cis.map_nominal(genotype_df, variant_df, phenotype_df, phenotype_pos_df, covariates_df, prefix)\n",
    "\n",
    "# genes on chr18\n",
    "cis.map_nominal(genotype_df, variant_df,\n",
    "                phenotype_df.loc[phenotype_pos_df['chr']=='chr18'],\n",
    "                phenotype_pos_df.loc[phenotype_pos_df['chr']=='chr18'],\n",
    "                prefix, covariates_df=covariates_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>maf</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_10644_C_G_b38</td>\n",
       "      <td>-98421</td>\n",
       "      <td>0.016854</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>0.580873</td>\n",
       "      <td>-0.117761</td>\n",
       "      <td>0.213125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_10847_C_A_b38</td>\n",
       "      <td>-98218</td>\n",
       "      <td>0.019101</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>0.142884</td>\n",
       "      <td>-0.298726</td>\n",
       "      <td>0.203505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_11275_G_A_b38</td>\n",
       "      <td>-97790</td>\n",
       "      <td>0.024719</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.745231</td>\n",
       "      <td>0.054619</td>\n",
       "      <td>0.167981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_11358_G_A_b38</td>\n",
       "      <td>-97707</td>\n",
       "      <td>0.024719</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0.745231</td>\n",
       "      <td>0.054619</td>\n",
       "      <td>0.167981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000263006.6</td>\n",
       "      <td>chr18_11445_G_A_b38</td>\n",
       "      <td>-97620</td>\n",
       "      <td>0.023596</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>0.603276</td>\n",
       "      <td>0.089378</td>\n",
       "      <td>0.171851</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        phenotype_id           variant_id  tss_distance       maf  ma_samples  \\\n",
       "0  ENSG00000263006.6  chr18_10644_C_G_b38        -98421  0.016854          15   \n",
       "1  ENSG00000263006.6  chr18_10847_C_A_b38        -98218  0.019101          17   \n",
       "2  ENSG00000263006.6  chr18_11275_G_A_b38        -97790  0.024719          22   \n",
       "3  ENSG00000263006.6  chr18_11358_G_A_b38        -97707  0.024719          22   \n",
       "4  ENSG00000263006.6  chr18_11445_G_A_b38        -97620  0.023596          21   \n",
       "\n",
       "   ma_count  pval_nominal     slope  slope_se  \n",
       "0        15      0.580873 -0.117761  0.213125  \n",
       "1        17      0.142884 -0.298726  0.203505  \n",
       "2        22      0.745231  0.054619  0.167981  \n",
       "3        22      0.745231  0.054619  0.167981  \n",
       "4        21      0.603276  0.089378  0.171851  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load results\n",
    "pairs_df = pd.read_parquet('{}.cis_qtl_pairs.chr18.parquet'.format(prefix))\n",
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *cis*-QTL: empirical p-values for phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cis-QTL mapping: empirical p-values for phenotypes\n",
      "  * 445 samples\n",
      "  * 301 phenotypes\n",
      "  * 26 covariates\n",
      "  * 13369268 variants\n",
      "  * using seed 123456\n",
      "  * computing permutations\n",
      "    processing phenotype 301/301\n",
      "  Time elapsed: 0.34 min\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# all genes\n",
    "# cis_df = cis.map_cis(genotype_df, variant_df, phenotype_df, phenotype_pos_df, covariates_df)\n",
    "\n",
    "# genes on chr18\n",
    "cis_df = cis.map_cis(genotype_df, variant_df, \n",
    "                     phenotype_df.loc[phenotype_pos_df['chr']=='chr18'],\n",
    "                     phenotype_pos_df.loc[phenotype_pos_df['chr']=='chr18'],\n",
    "                     covariates_df=covariates_df, seed=123456)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_var</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>true_df</th>\n",
       "      <th>pval_true_df</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>maf</th>\n",
       "      <th>ref_factor</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phenotype_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000263006.6</th>\n",
       "      <td>6120</td>\n",
       "      <td>1.038813</td>\n",
       "      <td>1138.435791</td>\n",
       "      <td>374.660431</td>\n",
       "      <td>8.220927e-40</td>\n",
       "      <td>chr18_112535_G_A_b38</td>\n",
       "      <td>3470</td>\n",
       "      <td>212</td>\n",
       "      <td>251</td>\n",
       "      <td>0.282022</td>\n",
       "      <td>1</td>\n",
       "      <td>4.050327e-44</td>\n",
       "      <td>0.726425</td>\n",
       "      <td>0.046171</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>3.677099e-38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000101557.14</th>\n",
       "      <td>6355</td>\n",
       "      <td>1.032236</td>\n",
       "      <td>1076.304199</td>\n",
       "      <td>370.176453</td>\n",
       "      <td>5.632799e-11</td>\n",
       "      <td>chr18_210698_T_C_b38</td>\n",
       "      <td>52315</td>\n",
       "      <td>192</td>\n",
       "      <td>222</td>\n",
       "      <td>0.249438</td>\n",
       "      <td>1</td>\n",
       "      <td>3.505407e-12</td>\n",
       "      <td>-0.191712</td>\n",
       "      <td>0.026749</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>3.498976e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000079134.11</th>\n",
       "      <td>6921</td>\n",
       "      <td>1.047217</td>\n",
       "      <td>1155.656616</td>\n",
       "      <td>370.356049</td>\n",
       "      <td>3.888747e-08</td>\n",
       "      <td>chr18_243547_T_A_b38</td>\n",
       "      <td>-24503</td>\n",
       "      <td>293</td>\n",
       "      <td>383</td>\n",
       "      <td>0.430337</td>\n",
       "      <td>1</td>\n",
       "      <td>5.473728e-09</td>\n",
       "      <td>-0.122720</td>\n",
       "      <td>0.020602</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>2.744039e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000263884.1</th>\n",
       "      <td>6921</td>\n",
       "      <td>1.039806</td>\n",
       "      <td>1152.500610</td>\n",
       "      <td>369.873535</td>\n",
       "      <td>7.681883e-04</td>\n",
       "      <td>chr18_584440_G_C_b38</td>\n",
       "      <td>316292</td>\n",
       "      <td>81</td>\n",
       "      <td>88</td>\n",
       "      <td>0.098876</td>\n",
       "      <td>1</td>\n",
       "      <td>3.540397e-04</td>\n",
       "      <td>-0.330811</td>\n",
       "      <td>0.091845</td>\n",
       "      <td>0.574843</td>\n",
       "      <td>5.695496e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000158270.11</th>\n",
       "      <td>8134</td>\n",
       "      <td>1.054920</td>\n",
       "      <td>1277.927734</td>\n",
       "      <td>369.469055</td>\n",
       "      <td>2.516541e-09</td>\n",
       "      <td>chr18_519222_C_T_b38</td>\n",
       "      <td>18500</td>\n",
       "      <td>108</td>\n",
       "      <td>115</td>\n",
       "      <td>0.129213</td>\n",
       "      <td>1</td>\n",
       "      <td>2.409727e-10</td>\n",
       "      <td>-0.388277</td>\n",
       "      <td>0.059808</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>1.567326e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    num_var  beta_shape1  beta_shape2     true_df  \\\n",
       "phenotype_id                                                        \n",
       "ENSG00000263006.6      6120     1.038813  1138.435791  374.660431   \n",
       "ENSG00000101557.14     6355     1.032236  1076.304199  370.176453   \n",
       "ENSG00000079134.11     6921     1.047217  1155.656616  370.356049   \n",
       "ENSG00000263884.1      6921     1.039806  1152.500610  369.873535   \n",
       "ENSG00000158270.11     8134     1.054920  1277.927734  369.469055   \n",
       "\n",
       "                    pval_true_df            variant_id  tss_distance  \\\n",
       "phenotype_id                                                           \n",
       "ENSG00000263006.6   8.220927e-40  chr18_112535_G_A_b38          3470   \n",
       "ENSG00000101557.14  5.632799e-11  chr18_210698_T_C_b38         52315   \n",
       "ENSG00000079134.11  3.888747e-08  chr18_243547_T_A_b38        -24503   \n",
       "ENSG00000263884.1   7.681883e-04  chr18_584440_G_C_b38        316292   \n",
       "ENSG00000158270.11  2.516541e-09  chr18_519222_C_T_b38         18500   \n",
       "\n",
       "                    ma_samples  ma_count       maf  ref_factor  pval_nominal  \\\n",
       "phenotype_id                                                                   \n",
       "ENSG00000263006.6          212       251  0.282022           1  4.050327e-44   \n",
       "ENSG00000101557.14         192       222  0.249438           1  3.505407e-12   \n",
       "ENSG00000079134.11         293       383  0.430337           1  5.473728e-09   \n",
       "ENSG00000263884.1           81        88  0.098876           1  3.540397e-04   \n",
       "ENSG00000158270.11         108       115  0.129213           1  2.409727e-10   \n",
       "\n",
       "                       slope  slope_se  pval_perm     pval_beta  \n",
       "phenotype_id                                                     \n",
       "ENSG00000263006.6   0.726425  0.046171   0.000100  3.677099e-38  \n",
       "ENSG00000101557.14 -0.191712  0.026749   0.000100  3.498976e-08  \n",
       "ENSG00000079134.11 -0.122720  0.020602   0.000100  2.744039e-05  \n",
       "ENSG00000263884.1  -0.330811  0.091845   0.574843  5.695496e-01  \n",
       "ENSG00000158270.11 -0.388277  0.059808   0.000100  1.567326e-06  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cis_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *trans*-QTL mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trans-QTL mapping\n",
      "  * 445 samples\n",
      "  * 19836 phenotypes\n",
      "  * 26 covariates\n",
      "  * 13369268 variants\n",
      "    processing batch 1337/1337\n",
      "    elapsed time: 0.67 min\n",
      "  * 7620376 variants passed MAF >= 0.05 filtering\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "# run mapping\n",
    "# to limit output size, only associations with p-value <= 1e-5 are returned\n",
    "trans_df = trans.map_trans(genotype_df, phenotype_df, covariates_df, batch_size=10000,\n",
    "                           return_sparse=True, pval_threshold=1e-5, maf_threshold=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove cis-associations\n",
    "trans_df = trans.filter_cis(trans_df, phenotype_pos_df.T.to_dict(), variant_df, window=5000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>maf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_10177_A_AC_b38</td>\n",
       "      <td>ENSG00000169203.16</td>\n",
       "      <td>6.147416e-06</td>\n",
       "      <td>0.352761</td>\n",
       "      <td>0.077021</td>\n",
       "      <td>0.428090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_30923_G_T_b38</td>\n",
       "      <td>ENSG00000278668.1</td>\n",
       "      <td>8.097087e-06</td>\n",
       "      <td>0.416201</td>\n",
       "      <td>0.092097</td>\n",
       "      <td>0.123595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_30923_G_T_b38</td>\n",
       "      <td>ENSG00000156531.16</td>\n",
       "      <td>1.868906e-06</td>\n",
       "      <td>-0.199628</td>\n",
       "      <td>0.041282</td>\n",
       "      <td>0.123595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_47159_T_C_b38</td>\n",
       "      <td>ENSG00000185324.21</td>\n",
       "      <td>8.467480e-25</td>\n",
       "      <td>-0.697666</td>\n",
       "      <td>0.063556</td>\n",
       "      <td>0.056180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1_49554_A_G_b38</td>\n",
       "      <td>ENSG00000271155.1</td>\n",
       "      <td>3.804473e-06</td>\n",
       "      <td>-0.452572</td>\n",
       "      <td>0.096611</td>\n",
       "      <td>0.051685</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            variant_id        phenotype_id          pval         b      b_se  \\\n",
       "0  chr1_10177_A_AC_b38  ENSG00000169203.16  6.147416e-06  0.352761  0.077021   \n",
       "2   chr1_30923_G_T_b38   ENSG00000278668.1  8.097087e-06  0.416201  0.092097   \n",
       "3   chr1_30923_G_T_b38  ENSG00000156531.16  1.868906e-06 -0.199628  0.041282   \n",
       "4   chr1_47159_T_C_b38  ENSG00000185324.21  8.467480e-25 -0.697666  0.063556   \n",
       "5   chr1_49554_A_G_b38   ENSG00000271155.1  3.804473e-06 -0.452572  0.096611   \n",
       "\n",
       "        maf  \n",
       "0  0.428090  \n",
       "2  0.123595  \n",
       "3  0.123595  \n",
       "4  0.056180  \n",
       "5  0.051685  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
